// @remix-project/remix-analyzer/solidity-analyzer/modules/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var txOrigin_1 = require("./txOrigin");
exports.txOrigin = txOrigin_1.default;
var gasCosts_1 = require("./gasCosts");
exports.gasCosts = gasCosts_1.default;
var thisLocal_1 = require("./thisLocal");
exports.thisLocal = thisLocal_1.default;
var checksEffectsInteraction_1 = require("./checksEffectsInteraction");
exports.checksEffectsInteraction = checksEffectsInteraction_1.default;
var constantFunctions_1 = require("./constantFunctions");
exports.constantFunctions = constantFunctions_1.default;
var similarVariableNames_1 = require("./similarVariableNames");
exports.similarVariableNames = similarVariableNames_1.default;
var inlineAssembly_1 = require("./inlineAssembly");
exports.inlineAssembly = inlineAssembly_1.default;
var blockTimestamp_1 = require("./blockTimestamp");
exports.blockTimestamp = blockTimestamp_1.default;
var lowLevelCalls_1 = require("./lowLevelCalls");
exports.lowLevelCalls = lowLevelCalls_1.default;
var blockBlockhash_1 = require("./blockBlockhash");
exports.blockBlockhash = blockBlockhash_1.default;
var noReturn_1 = require("./noReturn");
exports.noReturn = noReturn_1.default;
var selfdestruct_1 = require("./selfdestruct");
exports.selfdestruct = selfdestruct_1.default;
var guardConditions_1 = require("./guardConditions");
exports.guardConditions = guardConditions_1.default;
var deleteDynamicArrays_1 = require("./deleteDynamicArrays");
exports.deleteDynamicArrays = deleteDynamicArrays_1.default;
var assignAndCompare_1 = require("./assignAndCompare");
exports.assignAndCompare = assignAndCompare_1.default;
var erc20Decimals_1 = require("./erc20Decimals");
exports.erc20Decimals = erc20Decimals_1.default;
var stringBytesLength_1 = require("./stringBytesLength");
exports.stringBytesLength = stringBytesLength_1.default;
var intDivisionTruncate_1 = require("./intDivisionTruncate");
exports.intDivisionTruncate = intDivisionTruncate_1.default;
var etherTransferInLoop_1 = require("./etherTransferInLoop");
exports.etherTransferInLoop = etherTransferInLoop_1.default;
var deleteFromDynamicArray_1 = require("./deleteFromDynamicArray");
exports.deleteFromDynamicArray = deleteFromDynamicArray_1.default;
var forLoopIteratesOverDynamicArray_1 = require("./forLoopIteratesOverDynamicArray");
exports.forLoopIteratesOverDynamicArray = forLoopIteratesOverDynamicArray_1.default;
